[8.19] [AI4DSOC] [Attack discovery] Improves test coverage for Attack discovery alerts (#230682)

kibanamachine opened the pull request:
# Backport

This will backport the following commits from `main` to `8.19`:
 - [[AI4DSOC] [Attack discovery] Improves test coverage for Attack discovery alerts (#230682)](https://github.com/elastic/kibana/pull/230682)

<!--- Backport version: 9.6.6 -->

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--BACKPORT [{"author":{"name":"Andrew Macri","email":"andrew.macri@elastic.co"},"sourceCommit":{"committedDate":"2025-08-06T13:08:03Z","message":"[AI4DSOC] [Attack discovery] Improves test coverage for Attack discovery alerts (#230682)\n\n## [AI4DSOC] [Attack discovery] Improves test coverage for _Attack discovery alerts_\n\nThis PR improves unit test coverage for the [[AI4DSOC] [Attack discovery] Attack discovery alerts](https://github.com/elastic/kibana/pull/218906) feature.\n\n![attack_discovery_alerts](https://github.com/user-attachments/assets/7c69ceff-2a90-4d60-8792-51cdb87e62a2)\n\n_Above: The Attack discovery alerts feature_","sha":"545499701286c9bff29fd2e858a49fef88da94d5","branchLabelMapping":{"^v9.2.0$":"main","^v(\\d+).(\\d+).\\d+$":"$1.$2"}},"sourcePullRequest":{"labels":["release_note:skip","Team: SecuritySolution","Team:Security Generative AI","backport:version","v9.2.0","v9.1.1","v8.19.1"],"title":"[AI4DSOC] [Attack discovery] Improves test coverage for Attack discovery alerts","number":230682,"url":"https://github.com/elastic/kibana/pull/230682","mergeCommit":{"message":"[AI4DSOC] [Attack discovery] Improves test coverage for Attack discovery alerts (#230682)\n\n## [AI4DSOC] [Attack discovery] Improves test coverage for _Attack discovery alerts_\n\nThis PR improves unit test coverage for the [[AI4DSOC] [Attack discovery] Attack discovery alerts](https://github.com/elastic/kibana/pull/218906) feature.\n\n![attack_discovery_alerts](https://github.com/user-attachments/assets/7c69ceff-2a90-4d60-8792-51cdb87e62a2)\n\n_Above: The Attack discovery alerts feature_","sha":"545499701286c9bff29fd2e858a49fef88da94d5"}},"sourceBranch":"main","suggestedTargetBranches":["9.1","8.19"],"targetPullRequestStates":[{"branch":"main","label":"v9.2.0","branchLabelMappingKey":"^v9.2.0$","isSourceBranch":true,"state":"MERGED","url":"https://github.com/elastic/kibana/pull/230682","number":230682,"mergeCommit":{"message":"[AI4DSOC] [Attack discovery] Improves test coverage for Attack discovery alerts (#230682)\n\n## [AI4DSOC] [Attack discovery] Improves test coverage for _Attack discovery alerts_\n\nThis PR improves unit test coverage for the [[AI4DSOC] [Attack discovery] Attack discovery alerts](https://github.com/elastic/kibana/pull/218906) feature.\n\n![attack_discovery_alerts](https://github.com/user-attachments/assets/7c69ceff-2a90-4d60-8792-51cdb87e62a2)\n\n_Above: The Attack discovery alerts feature_","sha":"545499701286c9bff29fd2e858a49fef88da94d5"}},{"branch":"9.1","label":"v9.1.1","branchLabelMappingKey":"^v(\\d+).(\\d+).\\d+$","isSourceBranch":false,"state":"NOT_CREATED"},{"branch":"8.19","label":"v8.19.1","branchLabelMappingKey":"^v(\\d+).(\\d+).\\d+$","isSourceBranch":false,"state":"NOT_CREATED"}]}] BACKPORT-->

andrew-goldstein committed:
[AI4DSOC] [Attack discovery] Improves test coverage for Attack discovery alerts (#230682)

## [AI4DSOC] [Attack discovery] Improves test coverage for _Attack discovery alerts_

This PR improves unit test coverage for the [[AI4DSOC] [Attack discovery] Attack discovery alerts](https://github.com/elastic/kibana/pull/218906) feature.

![attack_discovery_alerts](https://github.com/user-attachments/assets/7c69ceff-2a90-4d60-8792-51cdb87e62a2)

_Above: The Attack discovery alerts feature_

(cherry picked from commit 545499701286c9bff29fd2e858a49fef88da94d5)

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/326998)
* Commit: 04fff0f66cf8d9599b8ccb711bcf3eca623a31da

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/326998#01987f8c-f2cf-452d-a8dd-5988636a4cc1)

cc @andrew-goldstein

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326998","number":326998,"commit":"04fff0f66cf8d9599b8ccb711bcf3eca623a31da"}],"number":326998}
buildkite-pr-comment-->

andrew-goldstein committed:
- fix 8.19 specific type errors

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/327138)
* Commit: 7c3e937cb1a4d9eca7ae338db6a3d9d82377a910
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-230785/7c3e937cb1a4d9eca7ae338db6a3d9d82377a910)

### Failed CI Steps
* [Jest Tests #17](https://buildkite.com/elastic/kibana-pull-request/builds/327138#0198804b-3049-4918-8687-34487fff939e)
* [Jest Tests #17](https://buildkite.com/elastic/kibana-pull-request/builds/327138#01988071-e705-4bef-a520-52c1c488b8f3)
* [FTR Configs #51](https://buildkite.com/elastic/kibana-pull-request/builds/327138#0198804b-3084-47ec-8f2d-4ab930893a08)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327138#0198804b-3049-4918-8687-34487fff939e) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327138/jobs/0198804b-3049-4918-8687-34487fff939e/artifacts/01988071-c3ca-4637-a764-320ce869081d) Jest Tests #<span></span>17 / combineFindAttackDiscoveryFilters when an alertIds filter is provided returns an alertIds filter for multiple alert IDs
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327138#01988071-e705-4bef-a520-52c1c488b8f3) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327138/jobs/01988071-e705-4bef-a520-52c1c488b8f3/artifacts/0198807d-16a3-4e31-9938-df0d96e8a081) Jest Tests #<span></span>17 / combineFindAttackDiscoveryFilters when an alertIds filter is provided returns an alertIds filter for multiple alert IDs
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327138#0198804b-3049-4918-8687-34487fff939e) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327138/jobs/0198804b-3049-4918-8687-34487fff939e/artifacts/01988071-c3ca-4a56-af23-42e9a95097a7) Jest Tests #<span></span>17 / combineFindAttackDiscoveryFilters when an alertIds filter is provided returns an alertIds filter for single alert ID
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327138#01988071-e705-4bef-a520-52c1c488b8f3) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327138/jobs/01988071-e705-4bef-a520-52c1c488b8f3/artifacts/0198807d-16a3-4d6c-9944-a65d347fe67f) Jest Tests #<span></span>17 / combineFindAttackDiscoveryFilters when an alertIds filter is provided returns an alertIds filter for single alert ID
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327138#0198804b-3049-4918-8687-34487fff939e) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327138/jobs/0198804b-3049-4918-8687-34487fff939e/artifacts/01988071-c3cb-4de3-a124-45d1410d5045) Jest Tests #<span></span>17 / combineFindAttackDiscoveryFilters when multiple filters are provided returns combined filters excluding empty arrays and undefined values
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327138#01988071-e705-4bef-a520-52c1c488b8f3) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327138/jobs/01988071-e705-4bef-a520-52c1c488b8f3/artifacts/0198807d-16a2-4b84-8576-6481acc3477a) Jest Tests #<span></span>17 / combineFindAttackDiscoveryFilters when multiple filters are provided returns combined filters excluding empty arrays and undefined values
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327138#0198804b-3049-4918-8687-34487fff939e) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327138/jobs/0198804b-3049-4918-8687-34487fff939e/artifacts/01988071-c3cb-4733-ba30-bae7ec0d6af9) Jest Tests #<span></span>17 / combineFindAttackDiscoveryFilters when multiple filters are provided returns combined filters with AND operator
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/327138#01988071-e705-4bef-a520-52c1c488b8f3) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/327138/jobs/01988071-e705-4bef-a520-52c1c488b8f3/artifacts/0198807d-16a3-445e-a68a-396bc9ca4787) Jest Tests #<span></span>17 / combineFindAttackDiscoveryFilters when multiple filters are provided returns combined filters with AND operator

### History
* :broken_heart: [Build #326998](https://buildkite.com/elastic/kibana-pull-request/builds/326998) failed 04fff0f66cf8d9599b8ccb711bcf3eca623a31da

cc @andrew-goldstein

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/327138","number":327138,"commit":"7c3e937cb1a4d9eca7ae338db6a3d9d82377a910"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/326998","number":326998,"commit":"04fff0f66cf8d9599b8ccb711bcf3eca623a31da"}],"number":327138}
buildkite-pr-comment-->

andrew-goldstein committed:
It appears that the following pull request <https://github.com/elastic/kibana/pull/222689> unintentionally merged a duplicate check (only) in the `8.19` branch:

```
    ...(alertIds && alertIds.length > 0
      ? [`(${alertIds.map((id) => `${ALERT_ATTACK_DISCOVERY_ALERT_IDS}: "${id}"`).join(' OR ')})`]
      : []),
```

as illustrated by the link below:

 https://github.com/elastic/kibana/blame/59dfa8fbb6d9a1a4f87603135d454c0769cb6b37/x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.ts#L64C3-L64C3

The duplicate check above is NOT present in the following branches:

- [main])(https://github.com/elastic/kibana/blame/main/x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.ts)
- [9.1](https://github.com/elastic/kibana/blame/9.1/x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.ts)

, which is why the [tests](https://github.com/elastic/kibana/blob/main/x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.test.ts) pass in `main` and `9.1`, but fail in the `8.19` branch with the following errors:

```
 FAIL  x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.test.ts
  combineFindAttackDiscoveryFilters
    when no filters are provided
      ✓ returns an empty string when all parameters are undefined (3 ms)
      ✓ returns an empty string when all arrays are empty
    when a search filter is provided
      ✓ returns the expected search filter for single search term
      ✓ returns a search filter with a trimmed search term
      ✓ returns an empty string when search is empty string
      ✓ returns an empty string when search is only whitespace
      ✓ returns a search filter with escaped special characters
    when an ids filter is provided
      ✓ returns ids filter for single id
      ✓ returns ids filter for multiple ids
      ✓ returns an empty string when ids is empty array
    when status filter is provided
      ✓ returns a status filter for single status
      ✓ returns a status filter for multiple statuses
      ✓ returns an empty string when status is empty array
    when a connectorNames filter is provided
      ✓ returns a connectorNames filter for single connector
      ✓ returns a connectorNames filter for multiple connectors
      ✓ returns an empty string when connectorNames is empty array
    when an alertIds filter is provided
      ✕ returns an alertIds filter for single alert ID (2 ms)
      ✕ returns an alertIds filter for multiple alert IDs
      ✓ returns an empty string when alertIds is empty array (1 ms)
    when a date range filters are provided
      ✓ returns a start date filter when start is provided
      ✓ returns an end date filter when end is provided
      ✓ returns both date filters when start and end are provided
    when multiple filters are provided
      ✕ returns combined filters with AND operator
      ✕ returns combined filters excluding empty arrays and undefined values (1 ms)
```

To make the currently failing tests (in the `8.19` branch) pass, the duplicate check will be removed from `x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.ts`,
which in-turn:

- makes the implementation the same across the `main`, `9.1`, and now `8.19` branches
- makes the tests pass ✅

andrew-goldstein commented on x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.ts:
```diff
@@ -61,10 +61,6 @@ export const combineFindAttackDiscoveryFilters = ({
             .join(' OR ')})`,
         ]
       : []),
-    ...(alertIds && alertIds.length > 0
-      ? [`(${alertIds.map((id) => `${ALERT_ATTACK_DISCOVERY_ALERT_IDS}: "${id}"`).join(' OR ')})`]
-      : []),
-
```
It appears that the following pull request <https://github.com/elastic/kibana/pull/222689> may have unintentionally merged a duplicate check (only) in the `8.19` branch:

```
    ...(alertIds && alertIds.length > 0
      ? [`(${alertIds.map((id) => `${ALERT_ATTACK_DISCOVERY_ALERT_IDS}: "${id}"`).join(' OR ')})`]
      : []),
```

as illustrated by the link below:

 https://github.com/elastic/kibana/blame/59dfa8fbb6d9a1a4f87603135d454c0769cb6b37/x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.ts#L64C3-L64C3

The duplicate check above is NOT present in the following branches:

- [main](https://github.com/elastic/kibana/blame/main/x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.ts)
- [9.1](https://github.com/elastic/kibana/blame/9.1/x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.ts)

, which is why the [tests](https://github.com/elastic/kibana/blob/main/x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.test.ts) pass in the `main` and `9.1`, but fail in the `8.19` branch with the following errors:

```
 FAIL  x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.test.ts
  combineFindAttackDiscoveryFilters
    when no filters are provided
      ✓ returns an empty string when all parameters are undefined (3 ms)
      ✓ returns an empty string when all arrays are empty
    when a search filter is provided
      ✓ returns the expected search filter for single search term
      ✓ returns a search filter with a trimmed search term
      ✓ returns an empty string when search is empty string
      ✓ returns an empty string when search is only whitespace
      ✓ returns a search filter with escaped special characters
    when an ids filter is provided
      ✓ returns ids filter for single id
      ✓ returns ids filter for multiple ids
      ✓ returns an empty string when ids is empty array
    when status filter is provided
      ✓ returns a status filter for single status
      ✓ returns a status filter for multiple statuses
      ✓ returns an empty string when status is empty array
    when a connectorNames filter is provided
      ✓ returns a connectorNames filter for single connector
      ✓ returns a connectorNames filter for multiple connectors
      ✓ returns an empty string when connectorNames is empty array
    when an alertIds filter is provided
      ✕ returns an alertIds filter for single alert ID (2 ms)
      ✕ returns an alertIds filter for multiple alert IDs
      ✓ returns an empty string when alertIds is empty array (1 ms)
    when a date range filters are provided
      ✓ returns a start date filter when start is provided
      ✓ returns an end date filter when end is provided
      ✓ returns both date filters when start and end are provided
    when multiple filters are provided
      ✕ returns combined filters with AND operator
      ✕ returns combined filters excluding empty arrays and undefined values (1 ms)
```

To make the currently failing tests (in the `8.19` branch) pass, the duplicate check will be removed from `x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/persistence/combine_find_attack_discovery_filters/index.ts`,
which in-turn:

- makes the implementation the same across the `main`, `9.1`, and now `8.19` branches
- makes the tests pass ✅
